# HW 01 (September 12th) - Avina Chan
**1) Picking dataset and importing into python notebook**

In [15]:
import pandas as pd

# URL of the dataset
url = 'https://raw.githubusercontent.com/mwaskom/seaborn-data/master/titanic.csv'

# Read the dataset into a pandas DataFrame
titanic_data = pd.read_csv(url)

# Display the first few rows of the dataset
titanic_data.head()

survived  pclass     sex   age  sibsp  parch     fare embarked  class  \
0         0       3    male  22.0      1      0   7.2500        S  Third   
1         1       1  female  38.0      1      0  71.2833        C  First   
2         1       3  female  26.0      0      0   7.9250        S  Third   
3         1       1  female  35.0      1      0  53.1000        S  First   
4         0       3    male  35.0      0      0   8.0500        S  Third   

     who  adult_male deck  embark_town alive  alone  
0    man        True  NaN  Southampton    no  False  
1  woman       False    C    Cherbourg   yes  False  
2  woman       False  NaN  Southampton   yes   True  
3  woman       False    C  Southampton   yes  False  
4    man        True  NaN  Southampton    no   True

**The code below was used to determine if dataset has missing values**

In [2]:
# Check for missing values in the dataset
missing_values = titanic_data.isnull().sum()

# Display the missing values
missing_values

survived         0
pclass           0
sex              0
age            177
sibsp            0
parch            0
fare             0
embarked         2
class            0
who              0
adult_male       0
deck           688
embark_town      2
alive            0
alone            0
dtype: int64

**2) Understanding the dataset size**

In [3]:
# Display the column names
columns = titanic_data.columns
print("Columns in the dataset:", columns)

# Get the shape of the dataset (number of rows and columns)
rows, cols = titanic_data.shape
print(f"\nThe dataset contains {rows} rows and {cols} columns.")

Columns in the dataset: Index(['survived', 'pclass', 'sex', 'age', 'sibsp', 'parch', 'fare',
       'embarked', 'class', 'who', 'adult_male', 'deck', 'embark_town',
       'alive', 'alone'],
      dtype='object')

The dataset contains 891 rows and 15 columns.


**Observations**: In statistics, **observations** refer to individual data points or records that contain attributes or features. This is equivalent to the number of rows in a dataset. For example, in the context of the Titanic dataset, each row represents a passenger on the Titanic, and the information collected for each passenger (such as 'survived', 'pclass', 'sex', 'age', etc.) is recorded in the columns and known as attributes or features. Therefore, each passenger is considered an **observation**.

**Variables**: As mentioned above, each observation contains attributes or features, which are known as **variables**. These **variables** correspond to the columns of the dataset, with each one containing information related to a specific aspect of the observations. In the context of the Titanic dataset, the **variables** would be 'survived', 'pclass', 'sex', 'age', etc. Additionally, there are two types of **variables** in the Titanic dataset: *Categorical (Qualitative) variables* (e.g., 'sex', 'pclass') and *Numerical (Quantitative) variables* (e.g.'age', 'fare').

**3) Summerizing Columns of the Dataset**

**Numerical Variables**

In [4]:
def summarize_data(df):
    print("Summary statistics for numerical variables:")
    print(df.describe(), '\n')

    #Call the function    
summarize_data(titanic_data)

Summary statistics for numerical variables:
         survived      pclass         age       sibsp       parch        fare
count  891.000000  891.000000  714.000000  891.000000  891.000000  891.000000
mean     0.383838    2.308642   29.699118    0.523008    0.381594   32.204208
std      0.486592    0.836071   14.526497    1.102743    0.806057   49.693429
min      0.000000    1.000000    0.420000    0.000000    0.000000    0.000000
25%      0.000000    2.000000   20.125000    0.000000    0.000000    7.910400
50%      0.000000    3.000000   28.000000    0.000000    0.000000   14.454200
75%      1.000000    3.000000   38.000000    1.000000    0.000000   31.000000
max      1.000000    3.000000   80.000000    8.000000    6.000000  512.329200 



**Categorial Variables**

In [5]:
def summarize_data(df):   
    print("Value counts for categorical variables:")
    
    #gives the count and the type of value 
    categorical_columns = df.select_dtypes(include='object').columns
    for column in categorical_columns:    
        print(df[column].value_counts())
        
# Call the function on the Titanic dataset
summarize_data(titanic_data)


Value counts for categorical variables:
sex
male      577
female    314
Name: count, dtype: int64
embarked
S    644
C    168
Q     77
Name: count, dtype: int64
class
Third     491
First     216
Second    184
Name: count, dtype: int64
who
man      537
woman    271
child     83
Name: count, dtype: int64
deck
C    59
B    47
D    33
E    32
A    15
F    13
G     4
Name: count, dtype: int64
embark_town
Southampton    644
Cherbourg      168
Queenstown      77
Name: count, dtype: int64
alive
no     549
yes    342
Name: count, dtype: int64


**4) df.shape v.s df.describe()**

**Discrepancy**  
It can be seen from the code above that when checking the dataset size (in section 2), the code (`df.shape`) concluded that there are 891 rows and 15 columns. However, when using `df.describe()`, it summarizes 6 numerical variables and 7 categorical variables, resulting in 2 missing columns.

**Explanation**  
This discrepancy can be explained by the fact that `df.describe()` summarizes numerical variables by default, so non-numerical variables are not shown, leading to only 6 variables being visible. Additionally, `df.describe(include='object')` is later used to summarize categorical columns, which only shows 7 variables. The missing two columns could be due to them not being strictly numerical or categorical. Examples of this include columns with mixed types or columns with all missing values. Additionally, there could be columns that don't fit into the numerical or categorical categories.

**5) Attribute vs Method**

**Attribute**  
An attribute does not contain `()`, as it is not a function; it is a **variable** that belongs to an object. It is typically used to get/set values related to the object. An example is `df.shape`, which is an attribute that provides the number of rows and columns of the dataset.

**Method**  
A method is a **function** that is associated with an object. Methods perform operations with the data within the object and return results. An example is `df.describe()`, which generates a summary of the statistics (mean, std, min, etc.) for numerical columns.

**Difference**  
Attributes are accessed directly and store information related to the object that can be modified later, as they are directly accessible and updatable. Methods, on the other hand, are functions that operate on data contained in the object. They perform calculations, data manipulations, or generate outputs. Attributes describe the state (what it is), whereas methods describe the behavior (what it can do).

**6) Understanding `df.describe()`** 

**count**: the number of non-missing observations for each variable <br>
**mean**: the arithmetic average of the value of each variable <br>
**std**: A measure of how spread out the values are from the average <br>
**Min**: The smallest value in the dataset for each variable <br>
**25%**: The value in which 25% of the data falls in (also known as the first quartile Q1) <br>
**50%**: The middle value of the data when it was ordered from smallest to largest (second quartile, Q2) <br>
**75%**: The value below which 75% of the data falls in (third quartile, Q3) <br>
**max**: The largest value in the dataset <br>

ChatGPT history: https://chatgpt.com/share/5da14d12-1962-44c7-8b12-896b25966246 <br>

Here's a summary of our interaction in this chat:

1. **Importing the Titanic Dataset**: 
   - You shared a URL for the Titanic dataset, and we discussed how to import it into a Python notebook.
   
2. **Analyzing Missing Values**: 
   - We identified missing values in the dataset using Pandas.
   
3. **Dataset Summary**:
   - I helped you understand the columns and the size of the dataset by using `df.shape`.
   
4. **Statistical Terminology**:
   - We discussed the term **"observations"** in the context of statistics, which refers to rows in the dataset, and **"variables"**, which are the columns or attributes being measured.
   - We also went over the types of variables in statistics.
   
5. **Basic Data Summarization**:
   - You asked for help with summarizing the Titanic dataset using `.describe()` and `.value_counts()`. I provided the necessary Python code for basic summarization using these methods.
   
6. **Discrepancy Between `df.shape` and `df.describe()`**:
   - We explored why there might be discrepancies between `df.shape` (showing 15 columns) and `df.describe()` (only summarizing some of the variables). I explained that some columns might be excluded based on their data types or due to missing values.
   
7. **Attributes vs Methods**:
   - We discussed the difference between attributes (e.g., `df.shape`) and methods (e.g., `df.describe()`), focusing on how attributes are used to access an object's state directly, while methods perform actions and require parentheses.

8. **Standard Deviation**:
   - I explained standard deviation in simpler terms, describing it as a measure of how spread out or close the data values are to the average.

9. **Value vs Object in Statistics**:
   - You asked whether "value" is also known as "object" in statistics, and I clarified the difference, explaining that a value is a specific data point, while an object may refer to the entity being observed.

Let me know if you need more details or further clarification on any of the topics we discussed!

**7) `df.dropna()` vs. `del df['col']`**

`df.dropna()`: thie **method** removes rows or column with missing values from the data, by default it will drop rows where at least one element is missing. Additionally you can customize this behaviour with parameters `df.dropna(axis=1)`(drop column), or `df.dropna(how = 'all')` (only drops rows/columns where all values are missing) <br>
<br>
`del df['col']`: command deletes a specific column from the data, regardless of whether the column contains missing values or not. This is only specific to columns. <br>
*the difference between the two is that `df.dropna()` is to remove rows or columns with missing data, whereas, `del df['col']` only deletes specific columns

1) **Example in which `df.dropna()` is preferred**  
An example is when you want to perform an analysis but need rows where all data points are complete (without any missing values). Therefore, the `df.dropna()` method is used.

**Reason this is preferred**  
- Helps preserve all relevant data by removing rows with missing values, ensuring that the analysis is based on complete information.

2) **Example in which `del df['col']` is preferred**  
An example of this is when you determine a specific column in your dataset is not relevant. Removing this column entirely can simplify your data and focus on the important features.

**Reason this is preferred**  
- It is easy to remove irrelevant data, which can streamline the dataset and improve analysis efficiency.

3) **Application order**  
It is important to apply `del df['col']` before `df.dropna()`. This is because removing unnecessary columns first allows `df.dropna()` to operate more efficiently by reducing the complexity of the dataset. By removing irrelevant columns, `df.dropna()` can focus on the remaining relevant data, improving its effectiveness in handling missing values.


In [10]:
clean_data = titanic_data.dropna()
clean_data
#this code assigns a new variable for the cleaned data 

survived  pclass     sex   age  sibsp  parch     fare embarked  class  \
1           1       1  female  38.0      1      0  71.2833        C  First   
3           1       1  female  35.0      1      0  53.1000        S  First   
6           0       1    male  54.0      0      0  51.8625        S  First   
10          1       3  female   4.0      1      1  16.7000        S  Third   
11          1       1  female  58.0      0      0  26.5500        S  First   
..        ...     ...     ...   ...    ...    ...      ...      ...    ...   
871         1       1  female  47.0      1      1  52.5542        S  First   
872         0       1    male  33.0      0      0   5.0000        S  First   
879         1       1  female  56.0      0      1  83.1583        C  First   
887         1       1  female  19.0      0      0  30.0000        S  First   
889         1       1    male  26.0      0      0  30.0000        C  First   

       who  adult_male deck  embark_town alive  alone  
1    woman       False    C    Cherbourg   yes  False  
3    woman       False    C  Southampton   yes  False  
6      man        True    E  Southampton    no   True  
10   child       False    G  Southampton   yes  False  
11   woman       False    C  Southampton   yes   True  
..     ...         ...  ...          ...   ...    ...  
871  woman       False    D  Southampton   yes  False  
872    man        True    B  Southampton    no   True  
879  woman       False    C    Cherbourg   yes  False  
887  woman       False    B  Southampton   yes   True  
889    man        True    C    Cherbourg   yes   True  

[182 rows x 15 columns]

In [11]:
clean_data.isnull().sum()
# this code checks if there are any missing values left in the data

survived       0
pclass         0
sex            0
age            0
sibsp          0
parch          0
fare           0
embarked       0
class          0
who            0
adult_male     0
deck           0
embark_town    0
alive          0
alone          0
dtype: int64

The reason I chose this approach is that I am unsure which data is considered unimportant. Therefore, it is only necessary to clean up the data with missing values. Before cleaning, the dataset contained 891 rows and 15 columns. After cleaning, it now contains 182 rows and 15 columns.

**Chat gpt log: https://chatgpt.com/share/95fb2af8-e1d2-478f-bbc2-9aecd0e14dfc**

Here's a summary of our chat:

1. **`df.dropna()` vs. `del df['col']`**:
   - **`df.dropna()`**: Removes rows (or columns) with missing values. Useful for cleaning data by eliminating incomplete records. It can be customized to drop rows or columns based on missing data.
   - **`del df['col']`**: Deletes a specific column from the DataFrame. Useful when a column is irrelevant or unnecessary for your analysis.

2. **Example for `df.dropna()`**:
   - Scenario: Removing rows with missing values in a dataset where complete data is required for analysis.

3. **Example for `del df['col']`**:
   - Scenario: Removing an irrelevant or redundant column from a DataFrame to simplify the dataset and focus on relevant features.

4. **When to use `del df['col']` before `df.dropna()`**:
   - **Simplification**: Removing unnecessary columns first to focus on relevant data.
   - **Efficiency**: Reducing the DataFrame’s size before handling missing values.
   - **Performance**: Improving processing time by dealing with a smaller, more focused DataFrame.

In summary, `del df['col']` is used to remove specific columns, while `df.dropna()` is used to handle missing values. Combining these methods can optimize data cleaning and analysis.

**8.1) Understanding `df.groupby()[].describe()`**

In [17]:
titanic_data.groupby("survived")["pclass"].describe()

count      mean       std  min  25%  50%  75%  max
survived                                                    
0         549.0  2.531876  0.735805  1.0  2.0  3.0  3.0  3.0
1         342.0  1.950292  0.863321  1.0  1.0  2.0  3.0  3.0

`df.groupby('col1')[col2].describe()` is a code snippet that provides a summary analysis of a column based on groups formed by another column. In the specific example provided below, `df.groupby('survived')` groups the data into two different categories (survived and not survived). By selecting `['pclass']`, the code summarizes the passenger class within the two groups defined by the 'survived' column. The data shows that the 'survived' column was divided into two groups (0 and 1), and a statistical summary of passenger classes for each group was generated. It can be noted that, on average, people who did not survive were more likely to be in higher classes, with a mean class of 2.53.

**8.2) `count` in `df.describe()` vs `df.groupby('col1')['col2'].describe()`**  
It is noted that the `count` section in `df.describe()` provides the number of values in the columns, offering a broad overview of the completeness of the entire dataset. In contrast, `df.groupby('col1')['col2'].describe()` shows the completeness of column `col2` within each of the subgroups defined in `col1`. This provides a better understanding of the completeness within each group and helps with assessing the reliability of the data within each subgroup. This is important because having some groups with significantly fewer data points than others can cause an imbalance that could impact the data distribution.

**Chat gpt logs: https://chatgpt.com/share/9a4fe1bf-4c31-4097-bef5-d5c2609e7307**

Certainly! Here’s a concise summary of our chat:

1. **Understanding `df.describe()` vs. `df.groupby("col1")["col2"].describe()`**:
   - **`df.describe()`**: Gives an overall summary of non-missing values and statistics for each column across the entire dataset.
   - **`df.groupby("col1")["col2"].describe()`**: Provides detailed statistics for `col2` within each subgroup defined by `col1`, including counts of non-missing values within each group.

2. **Key Differences**:
   - **Overall Completeness**: `df.describe()` reflects the completeness of each column globally.
   - **Group-Specific Completeness**: `df.groupby("col1")["col2"].describe()` shows how data completeness varies between different groups, which helps in understanding group-specific patterns and data distribution.

3. **Professor’s Point**:
   - The primary function of `count` with `groupby` is to reveal how data is distributed across groups and assess the representativeness of each group, beyond just addressing missing values.

4. **Implications**:
   - Understanding group-specific counts helps in assessing data reliability and making informed conclusions based on how data varies across different categories.

If you have any more questions or need further clarification, feel free to ask!

**8.3) Chatbot vs. Google**

**A) Forgot to include import pandas as pd in your code**

In [1]:
url = 'https://raw.githubusercontent.com/mwaskom/seaborn-data/master/titanic.csv'
titanic_data = pd.read_csv(url)

NameError: name 'pd' is not defined

For this problem, ChatGPT was able to provide an answer instantly, whereas using Google took longer and required scrolling through Stack Overflow.

**B) Typos (File not found error)**

In [6]:
import pandas as pd
url = 'titanics.csv'
titanic_data = pd.read_csv(url)

FileNotFoundError: [Errno 2] No such file or directory: 'titanics.csv'

In [5]:
import pandas as pd 
url = 'https://raw.githdfubusercosdfntent.com/mwaslom/segaborn-data/master/titanic.csv'
titanic_data = pd.read_csv(url)

URLError: <urlopen error [Errno -2] Name or service not known>

ChatGPT provided a faster and more accurate answer compared to Google searches. I found that using Google took me a long time to sift through people's answers and determine what was right.

**C) Dataframe before it's assigned into the variable**

In [7]:
import pandas as pd 
url = 'https://raw.githubusercontent.com/mwaskom/seaborn-data/master/titanic.csv'
titanic_data = pd.read_csv(url)
Titanic_Data.groupby("pclass")["survived"].describe()

NameError: name 'Titanic_Data' is not defined

ChatGPT was able to identify the issue immediately and provide a direct fix based on my previous chat logs about the Titanic dataset.

**D) Forgetting a Parentheses**

In [8]:
import pandas as pd 
url = 'https://raw.githubusercontent.com/mwaskom/seaborn-data/master/titanic.csv'
titanic_data = pd.read_csv(url

SyntaxError: incomplete input (2393863289.py, line 3)

Pasting the error into chat allowed for a quick answer that resolved the issue immediately.

**E) Mistyping the name of the chained functions with the code**

In [9]:
import pandas as pd 
url = 'https://raw.githubusercontent.com/mwaskom/seaborn-data/master/titanic.csv'
titanic_data = pd.read_csv(url)
titanic_data.group_by("age")["pclass"].describe()

AttributeError: 'DataFrame' object has no attribute 'group_by'


ChatGPT was also able to identify the error right away, noticing the `groupby` mistake.


**F) Using a column name that is not in the data**

In [11]:
titanic_data.groupby("hobbies")["age"].describe()

KeyError: 'hobbies'

Chatgpt was also able to identify this in its first attempt. 

**G) Forgetting to put the column name as a string**

In [12]:
titanic_data.groupby("sex")[age].describe()

NameError: name 'age' is not defined

Chat was able to identify this in its first attempt.

**ChatGPT logs: https://chatgpt.com/share/0f066c80-4b1d-4178-9cf1-b4370e9c15f8 <br>
<br>
Here’s a summary of our chat history:

1. **Importing and Reading Data**:
   - You encountered a `NameError` because `pandas` was not imported. Solution: `import pandas as pd`.
   - You faced a `FileNotFoundError` when trying to read a local file named `'titanics.csv'`. Solution: Ensure the file path is correct or use a valid URL.
   - You got a `URLError` when trying to fetch data from a URL, possibly due to network issues or restricted internet access. Solution: Check URL correctness or use a local file.

2. **Handling Syntax Errors**:
   - You had a `SyntaxError` due to an incomplete `pd.read_csv()` call. Solution: Ensure parentheses are correctly closed.

3. **Correct Usage of Methods**:
   - You experienced an `AttributeError` for using `group_by` instead of `groupby`. Solution: Use `groupby`.
   - You faced a `KeyError` for a column name `'hobbies'` that didn’t exist. Solution: Verify column names using `titanic_data.columns`.
   - You encountered a `NameError` due to using `age` without quotes. Solution: Enclose column names in quotes.

If you need any more details or further assistance, feel free to ask!

**9)** Yes, I interacted with a ChatBot and reviewed the course wiki-textbook to help with understanding all the material in the tutorial and lecture that i didnt understand